src/utils/charts/getChartData.js

```diff
@@ -2,175 +2,89 @@ import _ from 'lodash';
-  for (var incomei = 0; incomei < xRange.length; incomei++) {
-    var income = xRange[ incomei ];
-    clone.future.earned = income / multiplier;
-
-    // Datasets need to be in correct order for benefits to
-    // affect each other corrrectly.
-    for (let benefiti = 0; benefiti < benefitDatasets.length; benefiti++) {
-
-      let dataset = benefitDatasets[ benefiti ],
-          name    = dataset.benefitName,
-          funcs   = getBenefit[ name ];
-
-      let monthlyAmount = funcs.getSubsidy(clone, `future`);
-      dataset.data.push(monthlyAmount * multiplier);
-
-      // Mutate the clone in whatever way is appropriate for
-      // that program.
-      let newPropValues = funcs.getProps(clone, monthlyAmount);
-      Object.assign(clone.future, newPropValues);
-
-    }  // end for each benefit, in order
-
-  }  // end for all incomes
-
-  return;
-};  // End insertBenefitData()
-
-
-/** Returns the graph data formated in a way our graph library understands.
+ * @todo Change `activeBenefitsInOrder` to an array of
+ *     objects with a `name` prop that already contain
+ *     graph frosting? Shouldn't each graph take care
+ *     of its own frosting?
  * 
+ * @param {array} incomes
+ * @param {number} multiplier Number to convert from the
+ *     given time interval to monthly and back again -
+ *     benefit calculations are done with monthly values.
+ * @param {object} client Full client data, `current` and
+ *     `future` to be cloned, not mutated.
+ * @param {array} activeBenefitsInOrder Array of strings
+ *     of names of benefits in the order that we want them
+ *     to show up in the graph.
+ * @param {object} extraGraphProps Extra graph frosting on top of base frosting
  * @param {array} activeBenefits List of active benefits. @todo - use object instead?
  * */
-const getDatasets = function (xRange, client, multiplier, activeBenefits, extraProps) {
-
-  // Don't want to change actual client's data
-  var clone    = _.cloneDeep(client),
-      datasets = [];
-
-  for (let benefiti = 0; benefiti < benefitOrder.length; benefiti++) {
-
-    let benefitName = benefitOrder[ benefiti ];
-
-    if (activeBenefits.indexOf(benefitName) > -1) {
+const getChartData = function (incomes, multiplier, client, activeBenefitsInOrder, extraGraphProps) {
+  // Change client data to include object with array
+  // of benefits in the order we want?
+
+  // Should activeBenefits be an array of objects, each
+  // with their graph props and a name that's then used
+  // to get benefits' data in the right order?
+  // [{name: `income`, ...}, {name: `section8`, ...}]
+  // Existing `name` prop would be changed to `label`
+
+  // Active benefits should include 'income' if want it
+
+  // Shorter name now that we've clearly expressed what it is
+  var benefits = activeBenefitsInOrder;
+
+  var benefitDatasets = [],
+      allData         = {},  // each active benefit will have data in here
+      clone           = _.cloneDeep(client),
```

Comment:
Would it be better to re-clone the client separately on each iteration below, so that the benefit calculation starts afresh from the current situation for each income? Currently I don't think it makes a difference, but I could see a situation at some point where benefit A depends on the current value of benefit B (and how it affects expenses, etc.), so we wouldn't want benefit B's lingering value/side effects from a previous calculation to affect a second one. (And now my head hurts...)

Reply:
Hm OK, I see what you're saying and I want to agree with your points about why mutating `future` values should be safe, but I'm not fully convinced. For calculating SNAP `future` benefits for example, only the `future` client is used, so changes to `future` values could potentially affect that calculation. At the same, time, we need it to be able to see the change to future `rentShare` from the Section 8 calculation. 

I think it would be better if there were a way to guarantee that each benefit calculation only uses the `current` client, except for changes to future values that were made in the process of the current hypothetical situation. Cloning `future` each time is one way to do that. Or another way, and this is a radical proposal, could be to drop the idea of parallel `current` and `future` clients, and instead replace `future` with just a structure of all the differences from `current` in the future situation - call it `delta`. If `delta` doesn't have a certain key, that means it's the same as `current` (i.e.  the delta is implicitly 0). Then it would be a much smaller structure, and we could completely clear it before starting the benefit calculation process for a new hypothetical income.

Anyway that's clearly beyond the scope of this PR, and this should be merged!

As for the point about changing `current` values as we move forward or backward through incomes - I guess that depends on whether the benefit you get at a certain income depends on your previous income (or income history in general). If so, that makes things more complicated for sure...we'd have to decide whether the graph is meant to be showing what would happen if your income suddenly changed to each future income, or how benefits would change if your income slowly increased or decreased over time. In the latter case I think doing what you're saying makes sense. We should consider what effect it would have, though.